File: ios/FretTrainerEZ/Models/Note.swift
import Foundation

enum Note: Int, CaseIterable, Codable {
    case C = 0, Cs, D, Ds, E, F, Fs, G, Gs, A, As, B

    var sharpName: String {
        switch self {
        case .C: return "C"
        case .Cs: return "C#"
        case .D: return "D"
        case .Ds: return "D#"
        case .E: return "E"
        case .F: return "F"
        case .Fs: return "F#"
        case .G: return "G"
        case .Gs: return "G#"
        case .A: return "A"
        case .As: return "A#"
        case .B: return "B"
        }
    }

    var flatName: String {
        switch self {
        case .C: return "C"
        case .Cs: return "Db"
        case .D: return "D"
        case .Ds: return "Eb"
        case .E: return "E"
        case .F: return "F"
        case .Fs: return "Gb"
        case .G: return "G"
        case .Gs: return "Ab"
        case .A: return "A"
        case .As: return "Bb"
        case .B: return "B"
        }
    }

    func displayName(useFlats: Bool = false) -> String {
        useFlats ? flatName : sharpName
    }

    func advanced(by semitones: Int) -> Note {
        let raw = (self.rawValue + semitones % 12 + 12) % 12
        return Note(rawValue: raw)!
    }
}

File: android/app/src/main/kotlin/com/dontfretaboutitai/frettrainerez/models/Note.kt
package com.dontfretaboutitai.frettrainerez.models

enum class Note(val value: Int) {
    C(0), Cs(1), D(2), Ds(3), E(4), F(5), Fs(6), G(7), Gs(8), A(9), As(10), B(11);

    val sharpName: String get() = when (this) {
        C -> "C"; Cs -> "C#"; D -> "D"; Ds -> "D#"; E -> "E"; F -> "F"
        Fs -> "F#"; G -> "G"; Gs -> "G#"; A -> "A"; As -> "A#"; B -> "B"
    }

    val flatName: String get() = when (this) {
        C -> "C"; Cs -> "Db"; D -> "D"; Ds -> "Eb"; E -> "E"; F -> "F"
        Fs -> "Gb"; G -> "G"; Gs -> "Ab"; A -> "A"; As -> "Bb"; B -> "B"
    }

    fun displayName(useFlats: Boolean = false) = if (useFlats) flatName else sharpName

    fun advanced(by: Int): Note {
        val raw = ((value + by % 12) + 12) % 12
        return entries[raw]
    }

    companion object {
        val allCases: List<Note> get() = entries.toList()
    }
}
